### ISSUE TITLE
Broken types for `createApi`

### ISSUE THREAD

**Issue by vitonsky** [effector/effector#1048]
See [an example](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQQTMAGhXSzQFEA3NAOxiNU2wGUZo04BfOAMyghDgByNN25pkrKIIDcAKFnII1AM7xuECHAC8cADwAVOGgAe2agBNliDgD4AFAEptNxLICQilfAAkqttuJGNBY2AwBtAF17SIc5DyVVOAx8AIZSPGA7X0k0IgR3NwxzcztVUgAuOH1Iomo0AHcAMQ1K-Sd8t070GABXKGo4MIA6EbLsWobmiAi4tw4CAvM0ABs0bFKYCqqauCXV7HMp1vaCt26+gbG0Ie5gZewoOzswPlvVpy0XF4g39gBCLQ6PZrNCHDSxArzdwcCHuTyJdQQADyYBooNSJGwlBoMAM9ghbkRKLR5iG9SwyAAFnZksAhkVzLCzmsLq5Otk2AtOiN6fhoXIONIgA) of code
```ts
import { createApi, createEvent, createStore } from 'effector';

const foo = <T extends {}>() => {
	const $store = createStore<T[]>([]);
	const api = createApi($store, {
		add(state: T[], newFoo: T) {
			return [...state, newFoo];
		},
		delete(state: T[], deletedFoo: T) {
			return state.filter((profile) => profile !== deletedFoo);
		},
	});

	const fooOpened = createEvent<T>();
	fooOpened.watch(api.add); // types error here for `api.add`

	return {
		$store,
		...api
	};
};
```
![image](https://github.com/effector/effector/assets/86191922/0fd95d9f-f4a7-42bd-8205-93cee3532356)

# Current behavior

The code `fooOpened.watch(api.add);` emit unreadable type errorl

# Expected behavior

The code `fooOpened.watch(api.add);`  must no emit any errors, because event `fooOpened` receive the same type `T` as the `api.add`.

**Comment by zerobias** [comment]
Yes, looks like `createApi` types are complex enough to cause issues with generic types. Generic types are common source of issues in typings, so I suggest to use createEvent explicitly to avoid such issues until we fix that method

Also note that instead of `fooOpened.watch(api.add)` is better to use `sample({clock: fooOpened, target: api.add})`
https://withease.effector.dev/magazine/watch_calls.html
